content/en/blog/_posts/2021-07-21-memory-manager-beta.md

```diff
@@ -0,0 +1,106 @@
+---
+
+# Memory manager moves to beta
+
+The blog post describes the internals of the memory manager, a beta feature of Kubernetes 1.22. The **Memory Manager** is a component in the **kubelet** ecosystem proposed to enable the feature of guaranteed memory (and hugepages) allocation for pods in the [Guaranteed QoS class](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#qos-classes).
+
+This blog post covers:
+
+1. Why do you need it?
+2. The internal details of how the **MemoryManager** works
+3. Current limitations of the **MemoryManager**
+4. Future work for the **MemoryManager**
+
+## Why do you need it?
+
+To get the best performance and latency for the workload container, container CPUs, PCI devices, and memory should be aligned to the same NUMA node. The kubelet already provides a set of managers to align CPUs and PCI devices, but you did not have a way to align memory. By default, the kernel will try to allocate the memory from the same NUMA node where container CPUs are placed, but it can not guarantee it.
+
+## How does it work?
+
+The memory manager is doing two main things:
+- provides the topology hint to the topology manager
+- allocates the memory for containers and updates the state
+
+The overall sequence of the Memory Manager under the Kubelet
+
+![MemoryManagerDiagram
+](/images/blog/2021-07-21-memory-manager-beta/MemoryManagerDiagram.png "MemoryManagerDiagram")
+
+During the Admission phase:
+
+1. During the admission, the **Kubelet** calling the **TopologyManager** `Admit()` method.
+2. The **TopologyManager** is calling `GetTopologyHints()` for every hint provider including the **MemoryManager**.
+3. The **MemoryManager** calculates all possible NUMA nodes combinations for every container inside the pod and returns hints to the **TopologyManager**.
+4. The **TopologyManager** calls to `Allocate()` for every hint provider including the **MemoryManager**.
+5. The **MemoryManager** allocates the memory under the state according to the hint that the **TopologyManager** chose.
+
+During the creation phase:
+
+1. The **Kubelet** calls `PreCreateContainer()`.
+2. The **MemoryManager** get the NUMA nodes where it allocated the memory for the container and returns it to **Kubelet**.
+3. The **Kubelet** creates the container via CRI with the spec updated according to the **MemoryManager** information.
+
+### Let's talk about the configuration
+
+By default, the **MemoryManager** runs with the `None` policy, meaning it will just relax and will not do anything. To enable the **MemoryManager**, you should provide two **Kubelet** flags:
+
+- `--memory-manager-policy=Static`
+- `--reserved-memory="<numaNodeID>:<resourceName>=<quantity>"`
+
+It is clear what value you should set for the `--memory-manager-policy` parameter. But it does not 100% clear how you should configure the `--reserved-memory` parameter. To configure it correctly, you should follow two main rules:
```

Comment:
Change "it does not 100% clear" to "it is not 100% clear"

Reply:
done